src/os/bluestore/BlueStore.h

```diff
@@ -1086,276 +1086,281 @@ class BlueStore : public ObjectStore,
+      max = max_;
+    }
 
-    Cache(CephContext* cct) : cct(cct), logger(nullptr) {}
-    virtual ~Cache() {}
+    virtual void _trim_to(uint64_t max) = 0;
+    void _trim() {
+      _trim_to(max);
+    }
+    void trim() {
+      std::lock_guard l(lock);
+      _trim();    
+    }
+    void flush() {
+      std::lock_guard l(lock);
+     _trim_to(0);
+    }
 
-    virtual void _add_onode(OnodeRef& o, int level) = 0;
-    virtual void _rm_onode(OnodeRef& o) = 0;
-    virtual void _touch_onode(OnodeRef& o) = 0;
+#ifdef DEBUG_CACHE
+    virtual void _audit(const char *s) = 0;
+#else
+    void _audit(const char *s) { /* no-op */ }
+#endif
+  };
 
-    virtual void _add_buffer(Buffer *b, int level, Buffer *near) = 0;
-    virtual void _rm_buffer(Buffer *b) = 0;
-    virtual void _move_buffer(Cache *src, Buffer *b) = 0;
-    virtual void _adjust_buffer_size(Buffer *b, int64_t delta) = 0;
-    virtual void _touch_buffer(Buffer *b) = 0;
+  /// A Generic onode Cache Shard 
+  struct OnodeCacheShard : public CacheShard {
 
-    virtual uint64_t _get_num_onodes() = 0;
-    virtual uint64_t _get_buffer_bytes() = 0;
+  public:
+    OnodeCacheShard(CephContext* cct) : CacheShard(cct) {}
+    static OnodeCacheShard *create(CephContext* cct, string type,
+                                   PerfCounters *logger) {
+      OnodeCacheShard *c = nullptr;
+      // Currently we only implement an LRU cache for onodes
+      c = new LruOnodeCacheShard(cct);
+      c->logger = logger;
+      return c;
+    }
+
+    virtual void _add(OnodeRef& o, int level) = 0;
+    virtual void _rm(OnodeRef& o) = 0;
+    virtual void _touch(OnodeRef& o) = 0;
+    virtual uint64_t _get_num() = 0;
+    virtual void add_stats(uint64_t *onodes) = 0;
 
-    void add_extent() {
-      ++num_extents;
-    }
-    void rm_extent() {
-      --num_extents;
+    bool empty() {
+      std::lock_guard l(lock);
+      return _get_num() == 0;
     }
+  };
 
-    void add_blob() {
-      ++num_blobs;
+  /// An LRU onode Cache Shard
+  struct LruOnodeCacheShard : public OnodeCacheShard {
+    typedef boost::intrusive::list<
+      Onode,
+      boost::intrusive::member_hook<
+        Onode,
+        boost::intrusive::list_member_hook<>,
+        &Onode::lru_item> > list_t;
+    list_t lru;
+
+    LruOnodeCacheShard(CephContext *cct) : OnodeCacheShard(cct) {}
+
+    uint64_t _get_num() override {
+      return lru.size();
     }
-    void rm_blob() {
-      --num_blobs;
+    void _add(OnodeRef& o, int level) override {
+      (level > 0) ? lru.push_front(*o) : lru.push_back(*o);
     }
+    void _rm(OnodeRef& o) override {
+      lru.erase(lru.iterator_to(*o));
+    }
+    void _touch(OnodeRef& o) override {
+      lru.erase(lru.iterator_to(*o));
+      lru.push_front(*o);
+    }
+    void _trim_to(uint64_t max) override;
 
-    void set_onode_max(uint64_t max) {
-      onode_max = max;
+    void add_stats(uint64_t *onodes) override {
+      std::lock_guard l(lock);
+      *onodes += lru.size();
```

Comment:
may be start tracking  lru size using atomic_int and get rid off lock here?

Reply:
Yep, I agree.  Done!